### PR TITLE
[SPARK-14346] [SQL] Show create table 

### PR DESCRIPTION
**Pull request by xwu0226**
## What changes were proposed in this pull request?

Allow users to issue "`SHOW CREATE TABLE`" command natively in SparkSQL. 
-- For tables that are created by Hive, this command will display the DDL in hive syntax. If the syntax includes `CLUSTERED BY, SKEWED BY or STORED BY` clause, there will be a warning message saying that this DDL is not supported in SparkSQL native DDL yet. 

-- For tables that are created by datasource DDL, such as "`CREATE TABLE... USING ... OPTIONS (...)`", it will show the DDL in this syntax. 

-- For tables that are created by dataframe API, such as "`df.write.partitionBy(...).saveAsTable(...)`", currently the command will display DDL with the syntax "CREATE TABLE.. USING...OPTIONS(...)". However, this syntax lose the partitioning information. It is proposed to display create table in the dataframe API format (Will make this in another PR )
## How was this patch tested?

Unit tests are created.

### CODE REVIEW

**Comment by xwu0226** [comment]
@yhuai @andrewor14 I need to resubmit this PR and closed the earlier one because somehow my merge brought in a lot of other PR's stuff, which make the other PR un-reviewable.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 1]
```diff
@@ -423,6 +423,39 @@ case class ShowTablePropertiesCommand(
...
 
 /**
+ * A command for users to get the DDL of an existing table
+ * The syntax of using this command in SQL is:
+ * {{{
+ *   SHOW CREATE TABLE tableIdentifier
+ * }}}
+ */
+case class ShowCreateTableCommand(
+    tableName: String,
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Shouldn't we just pass in a `TableIdentifier`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -109,6 +109,15 @@ class SparkSqlAstBuilder extends AstBuilder {
   }
 
   /**
+   * Create a [[ShowCreateTableCommand]] logical plan
+   */
+  override def visitShowCreateTable(ctx: ShowCreateTableContext): LogicalPlan = withOrigin(ctx) {
+    ShowCreateTableCommand(
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Use `visitTableIdentifier`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 3]
```diff
@@ -625,11 +629,215 @@ private[hive] class HiveClientImpl(
...
+    if (schemaStringFromParts.isDefined) {
+      (schemaStringFromParts.map(s => DataType.fromJson(s).asInstanceOf[StructType]).
+        get map { f => s"${quoteIdentifier(f.name)} ${f.dataType.sql}" })
+        .mkString("( ", ", ", " )")
+    } else {
+      ""
+    }
+  }
+
+  private def generateDataSourceDDL(hiveTable: HiveTable): String = {
```
Comment (hvanhovell) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Can't we ask Hive to do this for us?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 1 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: @hvanhovell Thanks!. Yes, I should have. Will change it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 2 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: @hvanhovell Agree. Will change. Thanks!

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 3 above)
Comment (xwu0226) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @hvanhovell Are you referring to generateDataSourceDDL or generateHiveDDL?

Comment (xwu0226) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: @hvanhovell  I looked through the Hive's Hive, Table and MetaStoreClient APIs and have not found an API function that can do this for us. The way I see Hive implements this is through DDLTask.showCreateTable, which is executed through normal hive command parsing and executing route..

**Comment by gatorsmile** [comment]
CC @yhuai @andrewor14 The PR is ready for review. Thanks!

**Comment by wangmiao1981** [comment]
@xwu0226 use git rebase upstream/master. Do not use git merge upstream/master. I have the same issue before. git merge will add others' commits to your PR. git rebase will discard others' commits.

**Comment by xwu0226** [comment]
Thanks @wangmiao1981 .

This happens again. all the commits after my last commits got pulled into this PR. I need to close it and open a new PR. Will submit a new PR.

**Pull request closed without merging**